### PR TITLE
[UWP] Fix problem with async tasks not finishing before Auto size calculations are made

### PR DESCRIPTION
**Pull request by almedina-ms**
Fixes the crash where some parts of the auto image size calculations are made before the image has been loaded. I changed the auto sizing to be done after the image is loaded so we're sure the image source has been set and the properties are actually accesible.

Also, a resource resolver test was added to the test app so the resource resolver code is also checked when the test app is ran.

### CODE REVIEW

File: source/uwp/AdaptiveCardTestApp/ViewModels/RunningTestsViewModel.cs [hunk 1]
```diff
@@ -151,7 +153,11 @@ public async Task AddCardToTimeline(FileViewModel card)
 
         private async Task<TestResultViewModel> TestCard(FileViewModel cardFile, FileViewModel hostConfigFile)
         {
-            var renderResult = await RenderCard(cardFile, hostConfigFile);
+            Dictionary<string, IAdaptiveCardResourceResolver> resourceResolvers = new Dictionary<string, IAdaptiveCardResourceResolver>();
+            resourceResolvers.Add("symbol", new SampleResourceResolver());
+
+
```
Comment (paulcam206) [inline source/uwp/AdaptiveCardTestApp/ViewModels/RunningTestsViewModel.cs]: one `\n` too many :)

File: source/uwp/Renderer/lib/XamlBuilder.h [hunk 2]
```diff
@@ -144,8 +151,13 @@ namespace AdaptiveNamespace
         void SetImageOnUIElement(_In_ ABI::Windows::Foundation::IUriRuntimeClass* imageUrl,
                                  T* uiElement,
                                  ABI::AdaptiveNamespace::IAdaptiveCardResourceResolvers* resolvers,
+                                 bool isSizeAuto,
+                                 IInspectable* parentElement,
+                                 IInspectable* imageContainer,
                                  _Out_ bool* mustHideElement,
-                                 ABI::Windows::UI::Xaml::Media::Stretch stretch = Stretch_UniformToFill);
+                                 ABI::Windows::UI::Xaml::Media::Stretch stretch = Stretch_UniformToFill
+            );
```
Comment (paulcam206) [inline source/uwp/Renderer/lib/XamlBuilder.h]: is this what format document (Ctrl-K Ctrl-D) gave you?

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 3]
```diff
@@ -570,9 +573,9 @@ namespace AdaptiveNamespace
 
                 ComPtr<T> strongImageControl(uiElement);
                 ComPtr<XamlBuilder> strongThis(this);
-                THROW_IF_FAILED(getResourceStreamOperation->put_Completed(
+                HRESULT hr = (getResourceStreamOperation->put_Completed(
```
Comment (paulcam206) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >HRESULT hr = [](start = 16, length = 12)

why aren't we throwing here on failure any more?

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 4]
```diff
@@ -653,7 +663,7 @@ namespace AdaptiveNamespace
             ComPtr<XamlBuilder> strongThis(this);
             THROW_IF_FAILED(bufferWriteOperation->put_Completed(
                 Callback<Implements<RuntimeClassFlags<WinRtClassicComMix>, IAsyncOperationWithProgressCompletedHandler<UINT32, UINT32>>>(
-                    [strongThis, this, bitmapSource, randomAccessStream, strongImageControl](
+                    [strongThis, this, bitmapSource, randomAccessStream, strongImageControl, isAutoSize, parentElement, imageContainer /*ellipseAsShape, frameworkElement*/](
```
Comment (paulcam206) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >/*ellipseAsShape, frameworkElement*/ [](start = 135, length = 36)

remove

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 5]
```diff
@@ -1710,6 +1731,94 @@ namespace AdaptiveNamespace
...
+                    &eventToken));
+            }
+        }
+    }
+
+    _Use_decl_annotations_ template<typename T>
+    void XamlBuilder::SetAutoSize(T* destination, IInspectable* parentElement, IInspectable* imageContainer, bool mustHideElement)
+    {
+        ComPtr<IInspectable> parentElement2(parentElement);
+        //if container is ellipse
```
Comment (paulcam206) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >//if container is ellipse [](start = 8, length = 25)

is this accurate?

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 6]
```diff
@@ -1710,6 +1731,94 @@ namespace AdaptiveNamespace
...
+                    // what it would be otherwise, just set the visibility when we get the image
+                    return ellipseAsUIElement->put_Visibility(Visibility::Visibility_Visible);
+                })
+                    .Get(),
+                    &eventToken));
+            }
+        }
+    }
+
+    _Use_decl_annotations_ template<typename T>
```
Comment (paulcam206) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >_Use_decl_annotations_ [](start = 4, length = 22)

there aren't any declared annotations

Comment (paulcam206) [reply source/uwp/Renderer/lib/XamlBuilder.cpp]: (also above)

---
In reply to: [236882414](https://github.com/Microsoft/AdaptiveCards/pull/2191#discussion_r236882414) [](ancestors = 236882414)

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 7]
```diff
@@ -1710,6 +1731,94 @@ namespace AdaptiveNamespace
...
+    void XamlBuilder::SetAutoSize(T* destination, IInspectable* parentElement, IInspectable* imageContainer, bool mustHideElement)
+    {
+        ComPtr<IInspectable> parentElement2(parentElement);
```
Comment (paulcam206) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >ComPtr<IInspectable> parentElement2(parentElement); [](start = 8, length = 51)

this should be scoped inside the conditional block below. also, consider using a better name

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 8]
```diff
@@ -1710,6 +1731,94 @@ namespace AdaptiveNamespace
...
+        //if container is ellipse
+        if (parentElement != nullptr && m_enableXamlImageHandling)
+        {
+            ComPtr<IImage> xamlImage(destination);
```
Comment (paulcam206) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >ComPtr<IImage> xamlImage(destination); [](start = 12, length = 38)

declare this right before it's used

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 9]
```diff
@@ -1710,6 +1731,94 @@ namespace AdaptiveNamespace
...
+
+            // If the image hasn't loaded yet
+            if (mustHideElement)
+            {
+                // Collapse the Image control while the image loads, so that resizing is not noticeable
+                THROW_IF_FAILED(imageAsUIElement->put_Visibility(Visibility::Visibility_Collapsed));
+
+                // Handle ImageOpened event so we can check the imageSource's size to determine if it fits in its parent
+                EventRegistrationToken eventToken;
+                HRESULT hr = (xamlImage->add_ImageOpened(
```
Comment (paulcam206) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >HRESULT hr [](start = 16, length = 10)

are we doing anything with this? should we throw?

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 10]
```diff
@@ -1710,6 +1731,92 @@ namespace AdaptiveNamespace
         return S_OK;
     }
 
+    template<>
+    void XamlBuilder::SetAutoSize<IEllipse>(IEllipse* destination, IInspectable* parentElement, IInspectable* imageContainer, bool mustHideElement)
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >mustHideElement [](start = 131, length = 15)

Can we name this variable something more descriptive? It took me a bit to figure out what this represents. Maybe reverse it and call it IsImageLoaded.

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 11]
```diff
@@ -1710,6 +1731,92 @@ namespace AdaptiveNamespace
...
+            THROW_IF_FAILED(imageSource.As(&imageSourceAsBitmap));
+
+            // If the image hasn't loaded yet
+            if (mustHideElement)
+            {
+                // Collapse the Ellipse while the image loads, so that resizing is not noticeable
+                THROW_IF_FAILED(ellipseAsUIElement->put_Visibility(Visibility::Visibility_Collapsed));
+                // Handle ImageOpened event so we can check the imageSource's size to determine if it fits in its parent
+                EventRegistrationToken eventToken;
+                THROW_IF_FAILED(brushAsImageBrush->add_ImageOpened(
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >add_ImageOpened [](start = 51, length = 15)

Is it possible for the image to open before you set this handler?

Comment (almedina-ms) [reply source/uwp/Renderer/lib/XamlBuilder.cpp]: The problem we found was that the dataUri images and the images retrieved by resource resolvers wouldn't hit this event

---
In reply to: [239865550](https://github.com/Microsoft/AdaptiveCards/pull/2191#discussion_r239865550) [](ancestors = 239865550)

**Pull request merged**